view.c

```diff
@@ -164,98 +171,132 @@ Filerange view_viewport_get(View *view) {
 	return (Filerange){ .start = view->start, .end = view->end };
 }
 
-/* try to add another character to the view, return whether there was space left */
-static bool view_addch(View *view, Cell *cell) {
-	if (!view->line)
-		return false;
+static int view_max_text_width(const View *view) {
+	if (view->wrapcolumn > 0)
+		return MIN(view->wrapcolumn, view->width);
+	return view->width;
+}
 
-	int width;
-	size_t lineno = view->line->lineno;
-	unsigned char ch = (unsigned char)cell->data[0];
-	cell->style = view->cell_blank.style;
+static void view_wrap_line(View *view) {
+	int col = view->col;
+	int wrapcol = (view->wrapcol > 0) ? view->wrapcol : view->col;
+	Line *wrapped_line = view->line;
 
-	switch (ch) {
-	case '\t':
-		cell->width = 1;
-		width = view->tabwidth - (view->col % view->tabwidth);
-		for (int w = 0; w < width; w++) {
-			if (view->col + 1 > view->width) {
-				view->line = view->line->next;
-				view->col = 0;
-				if (!view->line)
-					return false;
-				view->line->lineno = lineno;
-			}
+	view->line = view->line->next;
+	view->col = 0;
+	view->wrapcol = 0;
 
-			cell->len = w == 0 ? 1 : 0;
-			int t = w == 0 ? SYNTAX_SYMBOL_TAB : SYNTAX_SYMBOL_TAB_FILL;
-			strncpy(cell->data, view->symbols[t]->symbol, sizeof(cell->data)-1);
-			view->line->cells[view->col] = *cell;
-			view->line->len += cell->len;
+	if (view->line) {
+		view->line->lineno = wrapped_line->lineno;
+		/* move extra cells to the next line */
+		for (int i = wrapcol; i < col; ++i) {
+			const Cell *cell = &wrapped_line->cells[i];
 			view->line->width += cell->width;
-			view->col++;
+			view->line->len += cell->len;
+			view->line->cells[view->col++] = *cell;
 		}
-		cell->len = 1;
-		return true;
-	case '\n':
-		cell->width = 1;
-		if (view->col + cell->width > view->width) {
-			view->line = view->line->next;
-			view->col = 0;
-			if (!view->line)
-				return false;
-			view->line->lineno = lineno;
+	}
+
+	/* clear remaining of line */
+	for (int i = wrapcol; i < view->width; ++i) {
+		if (i < col) {
+			wrapped_line->width -= wrapped_line->cells[i].width;
+			wrapped_line->len -= wrapped_line->cells[i].len;
 		}
+		wrapped_line->cells[i] = view->cell_blank;
+	}
+}
 
-		strncpy(cell->data, view->symbols[SYNTAX_SYMBOL_EOL]->symbol, sizeof(cell->data)-1);
+static bool view_add_cell(View *view, const Cell *cell) {
+	/* at most one iteration most of the time. we have to use a while-loop
+	 * here because of some pathological edge cases where one unicode char
+	 * may be bigger than the (extremely small) terminal width. */
```

Comment:
This still doesn't make sense to me. It can be either of "at most one
iteration" in which case the loop isn't needed OR "one iteration most
of the time" which you are suggesting is the case. But now the comment
makes it seem like a pointless endeavor because no one is using vis in
a terminal that is less than 1 (or even 10) character(s) wide.

Reply:
There are three cases:
1) Zero iterations: We are somewhere in the middle of the line. We have enough space to insert the next character and don't need to wrap the current line.
2) One iteration: Our terminal is 80 characters wide. We are inserting the 81st character. The while-loop runs once and wraps the current line. We can now insert another 80 (single width) characters until next line wrap.
3) Edge case: We have a one character wide terminal window and try to insert a double unicode. In this case the while-loop runs for every remaining line on the screen until no lines are left and we hit the `return false` branch. The editor obviously wont be able to render the text properly in this case, but it also won't crash. Calls to view_wrap_line in a loop make sure that the editor still does all the necessary housekeeping like setting the line number, replacing empty characters with empty cells and etc.

Sadly, I do not remember exactly what the problem was but it was something along those lines:

In my first version view_add_cell and view_wrap_line were implemented as mutually recursive functions. It didn't cause any problems in the most cases. However, sometimes after a line wrap there still wasn't enough space to insert the next character and the two functions were calling each other indefinitely. In aa54c376f3b9f36700a0e74d86fa871497c5e698 I have removed the recursion entirely and added a while-loop to account for the edge case of an extremely small terminal window.

Reply:
On second thought, I think it wasn't really the recursion itself causing the segfault, but rather a null pointer somewhere as a result of the mutual recursion. Anyway, no recursion -> no problems :)

Reply:
> recursion

Ahhh, I see now. I didn't catch that that was what you removed. If thats
the case it was probably running out of stack space. Let me think about
how to reword it a bit more and I will update the comment it before
merging.

Also that probably means that it wasn't an issue with your patch and
actually a preexisting issue. I'll make sure to keep it as a separate commit.

Reply:
> Also that probably means that it wasn't an issue with your patch
No, it was actually a problem I introduced in one of the first commits in this PR and not an preexisting issue. Both functions don't exist in master at all.

Feel free to reword my comments.

Reply:
Oh the `view_add_cell()` and `view_wrap_line()` mutually calling each
other. Maybe its ok to squash that then. Well I'll make sure whatever
I do makes sense before merging.

Reply:
So I did some more testing here and I'm not really able to recreate
the segfault even with a 1 ASCII char wide terminal and a file with
multiwide chars. I'm thinking we should just switch the `while` to an
`if` and not worry about the hypothetical edge case.

Reply:
It doesn't segfault anymore after the recursion got removed, but there is still an out-of-bounds write without the while-loop in `view->line->cells[view->col++] = cell_unused;`

We cannot replace `while` with `if` here.

Reply:
This doesn't fix the out-of-bounds write. When the terminal window is resized to be one character wide then `view->line->cells` will only have space for a single character.

This line will not produce an out-of-bounds write. `view->col` will be 0 after a resize. So this write is valid.
```
if (cell->width > 0)
	view->line->cells[view->col++] = *cell;
```

This for-loop is the problem in this case. If `cell->width` is for example two then we will write to `view->line->cells[1]` which is out-of-bounds for an array with only 1 element.
```
for (int i = 1; i < cell->width; i++)
	view->line->cells[view->col++] = cell_unused;
```

We should just keep the while-loop instead of removing it for doubtful gains and introducing new bugs in the process...

Reply:
Ok then the comment should explain this. Right now the code doesn't
make sense and the comment doesn't give any insight into what is going
on and just leads to confusion.

How about this:
```diff
 static bool view_add_cell(View *view, const Cell *cell) {
-	/* at most one iteration most of the time. we have to use a while-loop
-	 * here because of some pathological edge cases where one unicode char
-	 * may be bigger than the (extremely small) terminal width. */
+	/* if the terminal is resized to a single (ASCII) char an out
+	 * of bounds write could be performed for a multiwide char.
+	 * this can be caught by iterating through the lines with
+	 * view_wrap_line() until no lines remain.
+	 */
 	while (view->col + cell->width > view_max_text_width(view)) {
 		view_wrap_line(view);
 		if (!view->line)
```

Reply:
I am fine with that :+1: . I would just mention that "at most one iteration most of the time". It would make it clear that this is an inexpensive loop.